### ISSUE TITLE
How to build in Windows?

### ISSUE THREAD

**Issue by rubenmedrano** [AuLib/AuLib#1]
Hello Victor
I am not able to use it in codeblocks.
Do you know how can I build the library?

**Comment by vlazzarini** [comment]
you can run CMake targeting a Visual Studio project instead of Makefile, then load that project in Visual Studio.

Or use the Visual Studio CMake integration, looks like it's been supported for a couple of years:

https://devblogs.microsoft.com/cppblog/cmake-support-in-visual-studio/

Unfortunately, I don't use Windows so I can't help you any further.

Failing this, you could also make a Visual
Studio project by hand and add CPP files to it.

**Comment by rubenmedrano** [comment]
Thanks Victor for your quick response.
I am checking it and I hope I can make it work (I think the last option could be the one that works).
Best regards
Ruben

**Comment by rubenmedrano** [comment]
Hello again Victor

I was able to compile it by hand with codeblocks in Windows 10.
I had to modify one include in a few header files (SoundIn. and SoundOut.h) because thread doesn't workwith my config compiler in codeblocks. I changed the "include thread.h" by this: [https://github.com/meganz/mingw-std-threads](url) and it works.
As I have my own compiled "portaudio", "portmidi" and "libsndfile", I added it to the codeblocks project and also put this "defines": 

HAVE_PORTAUDIO
HAVE_PORTMIDI
HAVE_LIBSNDFILE

And it works. I have tested some examples without problems.
Thank you very much!

---------------------------------

The only problem I get is when I try to create the skeleton of an AudioBase-derived class, I get some errors:

There is a comma missing here (after "def_nchnls" and before "uint32_t vframes"):

    NewClass(double param = .5,
               uint32_t nchnls = def_nchnls uint32_t vframes = def_vframes,
               double sr = def_sr)

Here, "m_param(param)" should be "m_par(param)"?:
         `: m_param(param), AudioBase(nchnls, vframes, sr){};`



But there are other errors that I don't know how to fix. I don't know if I am in the wrong way to create the derived class.

For if it helps, what I've done is create a header file called "NewClass.h" and add as first line: 

`#include "AudioBase.h"`

and after goes your skeleton-example code.

And the compiler gives me the next errors:






||=== Build: Release in TestWx (compiler: GNU GCC Compiler) ===|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h||In constructor 'AuLib::NewClass::NewClass(double, uint32_t, uint32_t, double)':|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|11|warning: 'AuLib::NewClass::m_par' will be initialized after [-Wreorder]|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|17|warning:   base 'AuLib::AudioBase' [-Wreorder]|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|14|warning:   when initialized here [-Wreorder]|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h||In member function 'const double AuLib::NewClass::process(double)':|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|20|error: no matching function for call to 'AuLib::NewClass::dsp(const double&)'|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|8|note: candidate: virtual const double* AuLib::NewClass::dsp()|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|8|note:   candidate expects 0 arguments, 1 provided|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h||In member function 'const AuLib::NewClass& AuLib::NewClass::process(const AuLib::AudioBase&)':|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|30|error: invalid user-defined conversion from 'const double*' to 'const AuLib::AudioBase&' [-fpermissive]|

C:\Librerias\AuLib\AudioBase.h|39|note: candidate is: AuLib::AudioBase::AudioBase(uint32_t, uint32_t, double) <near match>|

C:\Librerias\AuLib\AudioBase.h|39|note:   conversion of argument 1 would be ill-formed:|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|30|error: invalid conversion from 'const double*' to 'uint32_t {aka unsigned int}' [-fpermissive]|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|30|error: invalid conversion from 'const double*' to 'uint32_t {aka unsigned int}' [-fpermissive]|

C:\Librerias\AuLib\AudioBase.h|39|note:   initializing argument 1 of 'AuLib::AudioBase::AudioBase(uint32_t, uint32_t, double)'|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|28|note:   initializing argument 1 of 'const AuLib::NewClass& AuLib::NewClass::process(const AuLib::AudioBase&)'|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|33|error: invalid initialization of reference of type 'const AuLib::NewClass&' from expression of type 'AuLib::NewClass*'|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h||In member function 'const AuLib::NewClass& AuLib::NewClass::process(const AuLib::AudioBase&, double)':|

X:\CodeBlocksProjects\TestWx 0.0.208 Aulib 03\NewClass.h|39|error: invalid initialization of reference of type 'const AuLib::NewClass&' from expression of type 'AuLib::NewClass*'|

||=== Build failed: 6 error(s), 3 warning(s) (0 minute(s), 4 second(s)) ===||`

**Comment by vlazzarini** [comment]
1) yes, m_param should be m_par
2) you need to define a dsp() method that
takes in a const double* as a parameter (an audio vector), and implement an audio
process of your own inside it.

The skeleton code is just a model for how the AudioBase class can be subclassed. It does not implement anything.

**Comment by rubenmedrano** [comment]
Yes, I was facing it right now. I am trying to create a new class for recording sound and storing it in a vector.
Thank you again!
